packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -1661,48 +1661,66 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
+
+            this.summarizerNode.startSummary(summaryRefSeqNum, summaryLogger);
+
+            // Helper function to check whether we should still continue between each async step.
+            const checkContinue = (): { continue: true; } | { continue: false; error: string } => {
+                // If summarizer loses connection it will never reconnect
+                if (!this.connected) {
+                    return { continue: false, error: "disconnected" };
+                }
+                // Ensure that lastSequenceNumber has not changed after pausing.
+                // We need the summary op's reference sequence number to match our summary sequence number,
+                // otherwise we'll get the wrong sequence number stamped on the summary's .protocol attributes.
+                if (this.deltaManager.lastSequenceNumber !== summaryRefSeqNum) {
+                    return {
+                        continue: false,
+                        // eslint-disable-next-line max-len
+                        error: `lastSequenceNumber changed before uploading to storage. ${this.deltaManager.lastSequenceNumber} !== ${summaryRefSeqNum}`,
+                    };
+                }
+                return { continue: true };
+            };
+
+            const attemptData: Omit<IUnsubmittedSummaryData, "error"> = {
                 referenceSequenceNumber: summaryRefSeqNum,
                 submitted: false,
             };
 
-            if (!this.connected) {
-                // If summarizer loses connection it will never reconnect
-                return { ...attemptData, reason: "disconnected" };
+            let continueResult = checkContinue();
+            if (!continueResult.continue) {
+                return { ...attemptData, error: continueResult.error };
             }
 
             const trace = Trace.start();
-            const summarizeResult = await this.summarize({
-                runGC: this.shouldRunGC,
-                fullTree,
-                trackState: true,
-                summaryLogger,
-            });
+            let summarizeResult: IChannelSummarizeResult;
+            try {
+                summarizeResult = await this.summarize({
+                    runGC: this.shouldRunGC,
+                    fullTree,
+                    trackState: true,
+                    summaryLogger,
+                });
+            } catch (error) {
+                return { ...attemptData, error };
+            }
 
             const generateData: IGeneratedSummaryData = {
                 summaryStats: summarizeResult.stats,
                 generateDuration: trace.trace().duration,
             };
 
-            if (!this.connected) {
-                return { ...attemptData, ...generateData, reason: "disconnected" };
+            continueResult = checkContinue();
+            if (!continueResult.continue) {
+                return { ...attemptData, ...generateData, error: continueResult.error };
```

Comment:
Signature of the function is not conveying that it returns extra props (IGeneratedSummaryData) when submitted === false.

Reply:
There are more props a bit below:
  return { ...attemptData, ...generateData, ...uploadData, error: continueResult.error };

Reply:
I think it does...
It returns either `IUnsubmittedSummaryData` or `ISubmittedSummaryData`, which are distinguishable by `submitted` as `false` or `true` respectively.

If `submitted === false`, then we can see `IUnsubmittedSummaryData` actually extends partials of both `IGeneratedSummaryData` and `IUploadedSUmmaryData`.

Reply:
Agree this isn't the easiest to read set of types, and will probably simplify them when using parameter object instead of return type (from other comment). But you can see definition of `IUnsubmittedSummaryData`:
```ts
export interface IUnsubmittedSummaryData extends Partial<IGeneratedSummaryData>, Partial<IUploadedSummaryData> { ... }
````